clients/src/main/java/org/apache/kafka/clients/producer/internals/BuiltInPartitioner.java

```diff
@@ -45,7 +45,7 @@ public class BuiltInPartitioner {
     private final AtomicReference<StickyPartitionInfo> stickyPartitionInfo = new AtomicReference<>();
 
     // Visible and used for testing only.
-    static volatile public Supplier<Integer> mockRandom = null;
+    protected static volatile Supplier<Integer> mockRandom = null;
```

Comment:
Maybe a better way is to add a package-private method `random()`, and then in the `BuiltInPartitionerTest`  we create a `BuiltInPartitioner` with override `random()`. With that change we can remove "mock" from `BuiltInPartitioner`.

Reply:
`mockRandom` is a static field, which means its behavior can be modified anywhere in the test. There are many instances in the context, not just within the test packages. Therefore, if we add a function and let the test module extend this class, it will cause the already created objects to not work as expected, requiring significant code changes.

Reply:
Agree that could introduce some changes, but we can minimize the changes by opening only required methods. For example:
**BuiltInPartitioner**
```java
    int random() {
        return Utils.toPositive(ThreadLocalRandom.current().nextInt());
    }

    private int nextPartition(Cluster cluster) {
        int random = random();
        ...
    }
```
**RecordAccumulator**
```java
    private Deque<ProducerBatch> getOrCreateDeque(TopicPartition tp) {
        TopicInfo topicInfo = topicInfoMap.computeIfAbsent(tp.topic(), k -> new TopicInfo(createBuiltInPartitioner(logContext, k, batchSize)));
        return topicInfo.batches.computeIfAbsent(tp.partition(), k -> new ArrayDeque<>());
    }
    BuiltInPartitioner createBuiltInPartitioner(LogContext logContext, String topic, int stickyBatchSize) {
        return new BuiltInPartitioner(logContext, topic, stickyBatchSize);
    }

        public TopicInfo(BuiltInPartitioner builtInPartitioner) {
            this.builtInPartitioner = builtInPartitioner;
        }
```

With those override-able methods, both `RecordAccumulatorTest` and `BuiltInPartitionerTest` can create `BuiltInPartitioner` with custom `random` method. WDYT?

Reply:
I modified this PR.
BTW,I moved the hard-coded parts to a private class within the test module in order to pass the original tests